diff -urN libtask_old/arm-ucontext.h libtask/arm-ucontext.h
--- libtask_old/arm-ucontext.h	1970-01-01 01:00:00.000000000 +0100
+++ libtask/arm-ucontext.h	2014-03-06 19:24:21.578125000 +0000
@@ -0,0 +1,64 @@
+typedef struct mcontext mcontext_t;
+typedef struct ucontext ucontext_t;
+
+extern	int		swapcontext(ucontext_t*, const ucontext_t*);
+extern	void	makecontext(ucontext_t*, void(*)(), int, ...);
+
+/*
+ * Copyright (c) 1992, 1993
+ *	The Regents of the University of California.  All rights reserved.
+ *
+ * This code is derived from software contributed to Berkeley by
+ * Ralph Campbell.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 4. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ *	@(#)ucontext.h	8.1 (Berkeley) 6/10/93
+ *	JNPR: ucontext.h,v 1.2 2007/08/09 11:23:32 katta
+ * $FreeBSD: src/sys/mips/include/ucontext.h,v 1.2 2010/01/10 19:50:24 imp Exp $
+ */
+
+
+struct mcontext {
+	int gregs[16];
+};
+
+struct ucontext {
+	/*
+	 * Keep the order of the first two fields. Also,
+	 * keep them the first two fields in the structure.
+	 * This way we can have a union with struct
+	 * sigcontext and ucontext_t. This allows us to
+	 * support them both at the same time.
+	 * note: the union is not defined, though.
+	 */
+	sigset_t	uc_sigmask;
+	mcontext_t	uc_mcontext;
+
+	struct __ucontext *uc_link;
+	stack_t		uc_stack;
+	int		__spare__[8];
+};
+
diff -urN libtask_old/asm.S libtask/asm.S
--- libtask_old/asm.S	2011-01-02 23:43:14.000000000 +0000
+++ libtask/asm.S	2014-04-02 12:47:31.218750000 +0100
@@ -1,5 +1,11 @@
 /* Copyright (c) 2005-2006 Russ Cox, MIT; see COPYRIGHT */
 
+#if defined (__arm__)
+#define NEEDARMCONTEXT 1
+#define SET setmcontext
+#define GET getmcontext
+#endif
+
 #if defined(__FreeBSD__) && defined(__i386__) && __FreeBSD__ < 5
 #define NEEDX86CONTEXT 1
 #define SET setmcontext
diff -urN libtask_old/context.c libtask/context.c
--- libtask_old/context.c	2011-01-02 23:43:14.000000000 +0000
+++ libtask/context.c	2014-04-02 12:43:59.609375000 +0100
@@ -2,6 +2,12 @@
 
 #include "taskimpl.h"
 
+
+#if defined (__arm__)
+#define NEEDARMMAKECONTEXT
+#define NEEDSWAPCONTEXT
+#endif
+
 #if defined(__APPLE__)
 #if defined(__i386__)
 #define NEEDX86MAKECONTEXT
diff -urN libtask_old/Makefile libtask/Makefile
--- libtask_old/Makefile	2009-10-10 05:55:25.000000000 +0100
+++ libtask/Makefile	2014-04-02 13:44:14.250000000 +0100
@@ -6,19 +6,17 @@
 	$(ASM)\
 	channel.o\
 	context.o\
-	fd.o\
-	net.o\
 	print.o\
 	qlock.o\
 	rendez.o\
 	task.o\
 
-all: $(LIB) primes tcpproxy testdelay
+all: $(LIB)
 
 $(OFILES): taskimpl.h task.h 386-ucontext.h power-ucontext.h
 
-AS=gcc -c
-CC=gcc
+AS=arm-eabi-gcc -c
+CC=arm-eabi-gcc
 CFLAGS=-Wall -c -I. -ggdb
 
 %.o: %.S
diff -urN libtask_old/task.c libtask/task.c
--- libtask_old/task.c	2009-10-10 05:33:49.000000000 +0100
+++ libtask/task.c	2014-04-02 12:48:03.296875000 +0100
@@ -98,8 +98,8 @@
 
 	/* do a reasonable initialization */
 	memset(&t->context.uc, 0, sizeof t->context.uc);
-	sigemptyset(&zero);
-	sigprocmask(SIG_BLOCK, &zero, &t->context.uc.uc_sigmask);
+//	sigemptyset(&zero);
+//	sigprocmask(SIG_BLOCK, &zero, &t->context.uc.uc_sigmask);
 
 	/* must initialize with current context */
 	if(getcontext(&t->context.uc) < 0){
@@ -352,17 +352,6 @@
 int
 main(int argc, char **argv)
 {
-	struct sigaction sa, osa;
-
-	memset(&sa, 0, sizeof sa);
-	sa.sa_handler = taskinfo;
-	sa.sa_flags = SA_RESTART;
-	sigaction(SIGQUIT, &sa, &osa);
-
-#ifdef SIGINFO
-	sigaction(SIGINFO, &sa, &osa);
-#endif
-
 	argv0 = argv[0];
 	taskargc = argc;
 	taskargv = argv;
@@ -371,8 +360,6 @@
 		mainstacksize = 256*1024;
 	taskcreate(taskmainstart, nil, mainstacksize);
 	taskscheduler();
-	fprint(2, "taskscheduler returned in main!\n");
-	abort();
 	return 0;
 }
 
diff -urN libtask_old/taskimpl.h libtask/taskimpl.h
--- libtask_old/taskimpl.h	2011-01-02 23:43:14.000000000 +0000
+++ libtask/taskimpl.h	2014-04-02 12:34:44.531250000 +0100
@@ -11,6 +11,12 @@
 
 #define USE_UCONTEXT 1
 
+#if defined(__arm__)
+#undef USE_UCONTEXT
+#define USE_UCONTEXT 0
+#endif
+
+
 #if defined(__OpenBSD__) || defined(__mips__)
 #undef USE_UCONTEXT
 #define USE_UCONTEXT 0
@@ -38,7 +44,11 @@
 #if USE_UCONTEXT
 #include <ucontext.h>
 #endif
+
+#if !defined(__arm__)
 #include <sys/utsname.h>
+#endif
+
 #include <inttypes.h>
 #include "task.h"
 
@@ -124,6 +134,7 @@
 #endif
 
 #if defined(__arm__)
+#include "arm-ucontext.h"
 int getmcontext(mcontext_t*);
 void setmcontext(const mcontext_t*);
 #define	setcontext(u)	setmcontext(&(u)->uc_mcontext)
